### ISSUE TITLE
Add Two Factor Authentication to Cloud Service

### ISSUE THREAD

**Issue by Ebreium** [klinker-apps/messenger-issues#667]
- [✅] I have [searched](https://github.com/klinker-apps/messenger-issues/issues), by keyword, for my issue/request to verify that there aren't any open or closed issues around this request.

**Request:**
To enhance the security of the Pulse message syncing service, I am suggesting that two factor authentication be added which supports the TOTP algorithm implemented by authentication apps such as Google Authenticator and [Authy](https://authy.com/).
By requiring a user to enter both their correct password and a code generated by their two factor authentication app, the risk of a user's messages being acquired by a malicious party due to their password being compromised would be mitigated.

**Comment by codex704** [comment]
That would be epic

**Comment by tnyeanderson** [comment]
Any updates on this? Would love to lock it down a bit more!

**Comment by klinker24** [comment]
If I haven't closed it, then it is planned for the future. If I had updates, I would post them in this topic.

**Comment by BlinkersFr31** [comment]
It's one of the most important future for me

**Comment by csm10495** [comment]
Anything we can do to accelerate this? I'm nervous that something so close to home isn't protected in this day and age of password issues, hacks, etc.

**Comment by abpages** [comment]
Yes this is crucial it's added. If they gain access to our sms via the web then effectively they have access to any other services that use 2 factor auth and send an sms to our phones. ....... It's a major security hole. It needs to have 2FA asap.

**Comment by csm10495** [comment]
@klinker24 any update (or planning) on this? Are there tasks the OS community can help with to make this happen faster?

**Comment by klinker24** [comment]
Someone could intercept your SMS without my app. SMS isn't secure to begin with and, if you are security-conscious, it shouldn't be an assumption that 2FA codes are safe through SMS.

I don't have any update or planning around this, sorry. Every part of the app is open source, other than the backend. If you wanted to contribute, feel free to make the front-end implementations for all of the different platforms (android, ios, web, tizen). There would need to be systems in place to turn on this functionality, to verify a users phone number, and to insert itself into the login process. There is plenty of work that needs done on this, without touching the backend

**Comment by tnyeanderson** [comment]
@klinker24 my account had its password reset (I guess) today by (I guess) some malicious party. I've had the same login for a long time and it worked until today, so unless there was some crazy backend auth issue going on..... then my account got compromised. I had to refresh the account with a different email (can't use the same one), then delete the original account, then refresh again with the same email. Hoping that all my messages, etc will be restored, which it appears they will... for now

Can we *please* take this as a sign that 2FA is important? It is really standard everywhere, and it should be implemented for **any** web-connected service, much less one that stores all my text messages! Who knows what the attacker got of mine because there was no 2FA??

Really hoping this is more heavily considered soon. Thank you for the hard work that you do for open source and people like myself. I will be a lifetime pulse user as long as it is around..... but this is a serious problem.

**Comment by klinker24** [comment]
Your account didn't get compromised. Pulse had 20 mins of downtime this morning.

**Comment by tnyeanderson** [comment]
Thanks Luke, I've updated my comment so it isn't misleading.

**Comment by tnyeanderson** [comment]
@klinker24 is there a downtime monitor I can follow for Pulse so I can avoid going through these hoops in the future? I checked the klinker twitter and didn't see anything about downtime...
